### ISSUE TITLE
"ShowMissingWorkloads" task failed unexpectedly.

### ISSUE THREAD

**Issue by PureWeen** [dotnet/sdk#19158]
I'm not sure the best steps to reproduce this issue but I was recently in a state where all my workloads (except MAUI) were in preview 7. The Maui workload was still in preview6.

When I tried to build any MAUI applications I would get the following exception
```
/usr/local/share/dotnet/sdk/6.0.100-preview.7.21327.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.ImportWorkloads.targets(33,5): error MSB4018: The "ShowMissingWorkloads" task failed unexpectedly. [/Users/shaneneuville/Projects/maui/src/Controls/src/Core/Controls.Core-net6.csproj]
```

@jonathanpeppers was telling me if a workload is missing it should tell me what workload is missing but it seems like the task that should be telling me this was failing. 

I've attached a binlog that will hopefully reveal something useful 

Here are the versions of things I had installed when I was seeing this exception
```
  ✔ android-aot (Microsoft.NET.Sdk.Android.Manifest-6.0.100 : 30.0.100-preview.7.91) installed.
  ✔ ios (Microsoft.NET.Sdk.iOS.Manifest-6.0.100 : 15.0.100-preview.7182) installed.
  ✔ maccatalyst (Microsoft.NET.Sdk.MacCatalyst.Manifest-6.0.100 : 15.0.100-preview.7182) installed.
  ✔ tvos (Microsoft.NET.Sdk.tvOS.Manifest-6.0.100 : 15.0.100-preview.7182) installed.
  ✔ macos (Microsoft.NET.Sdk.macOS.Manifest-6.0.100 : 12.0.100-preview.7182) installed.
  ✔ maui (Microsoft.NET.Sdk.Maui.Manifest-6.0.100 : 6.0.100-preview.6.1001) installed.
  ✔ microsoft-net-runtime-android (microsoft.net.workload.mono.toolchain.manifest-6.0.100 : 6.0.0-preview.6.21352.12) installed.
```

### BinLog
[ShowMissingWorkloadsCrash.zip](https://github.com/dotnet/sdk/files/6863337/ShowMissingWorkloadsCrash.zip)

**Comment by wli3** [comment]
@PureWeen All workload name changed between preview 6 and preview 7. Considering the error is about missing workload, it is very likely caused by that. And we don't try to maintain backcompact during previews. Could you try to use all preview 7 workload?

**Comment by PureWeen** [comment]
@wli3 I haven't seen this error again and I'm not really sure how I would currently go about reproducing to see if this is fixed or not.

It sounds like this was just a quirk of being in preview so I'll close for now and then re-open if I see the issue again.

**Comment by bakerhillpins** [comment]
Google gets 3 hits for "ShowMissingWorkloads" so I'm hoping this might be useful. I see this was closed without resolution or further information so I thought I'd add my experience as I'm seeing this after an upgrade of Visual Studio.  I'm going to be actively trying to correct this situation and if/when I stumble on a change that corrects it I'll comment. This is my experience so far.

- I have VS 2022 pro and pro preview installed. 
- I was actively developing several MAUI based applications with versions 17.3.6 and 17.4.0 preview 6 installed.
- I typically do my development work in the released VS (Non Preview) version 17.3.6. So I was actively building and running on 17.3.6 2 days ago.
- I updated BOTH Visual Studio versions yesterday AM. Updated Pro first,17.4.0 from 17.3.6. Then to Pro Preview version 17.5.0 preview 1 from 17.4.0 preview 6.
- Rebooted
- Opened my existing .NET 6 MAUI app solutions and could not build anything. Get a pile of errors, multiples of the following:

```xml
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB4018	The "ShowMissingWorkloads" task failed unexpectedly.
Microsoft.NET.Sdk.WorkloadManifestReader.WorkloadManifestCompositionException: Could not find workload 'ios' extended by workload 'maui-ios' in manifest 'microsoft.net.sdk.maui' [C:\Program Files\dotnet\sdk-manifests\6.0.300\microsoft.net.sdk.maui\WorkloadManifest.json]
   at Microsoft.NET.Sdk.WorkloadManifestReader.WorkloadResolver.<>c__DisplayClass24_0.<<EnumerateWorkloadWithExtends>g__EnumerateWorkloadWithExtendsRec|0>d.MoveNext()
   at Microsoft.NET.Sdk.WorkloadManifestReader.WorkloadResolver.<>c__DisplayClass24_0.<<EnumerateWorkloadWithExtends>g__EnumerateWorkloadWithExtendsRec|0>d.MoveNext()
   at Microsoft.NET.Sdk.WorkloadManifestReader.WorkloadResolver.<>c__DisplayClass24_0.<<EnumerateWorkloadWithExtends>g__EnumerateWorkloadWithExtendsRec|0>d.MoveNext()
   at System.Linq.Enumerable.All[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.NET.Sdk.WorkloadManifestReader.WorkloadResolver.<GetAvailableWorkloadDefinitions>d__29.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.NET.Sdk.WorkloadManifestReader.WorkloadResolver.GetWorkloadSuggestionForMissingPacks(IList`1 packIds, ISet`1& unsatisfiablePacks)
   at Microsoft.NET.Build.Tasks.ShowMissingWorkloads.ExecuteCore()
   at Microsoft.NET.Build.Tasks.TaskBase.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()	xxx.About	C:\Program Files\dotnet\sdk\6.0.403\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.ImportWorkloads.targets	38	
Error	NU1008	Projects that use central package version management should not define the version on the PackageReference items but on the PackageVersion items: DotNet.ReproducibleBuilds;Nerdbank.GitVersioning.	xxxxx.About	C:\Repos\d\yyyy\xxx.About\xxx.About.csproj	1	
Error	NETSDK1084	There is no application host available for the specified RuntimeIdentifier 'iossimulator-x64'.	xxx.Teammate	C:\Program Files\dotnet\sdk\6.0.403\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets	131	
Error	NU1012	Platform version is not present for one or more target frameworks, even though they have specified a platform: net6.0-android, net6.0-ios, net6.0-maccatalyst	xxx.Teammate	C:\Repos\d\yyyy\xxx.Teammate\xxx.Teammate\xxx.Teammate.csproj	1	
Severity	Code	Description	Project	File	Line	Suppression State
Error	XA5207	Could not find android.jar for API level 31. This means the Android SDK platform for API level 31 is not installed. Either install it in the Android SDK Manager (Tools > Android > Android SDK Manager...), or change the Xamarin.Android project to target an API version that is installed. (C:\Program Files (x86)\Android\android-sdk\platforms\android-31\android.jar missing.)	xxx.Teammate	C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\32.0.476\tools\Xamarin.Android.Common.targets	533	
```

- Used VS Pro to create a NOOB MAUI app (MauiApp1). Selected .NET 6 as the target. It also does not build. 
- Used VS Pro to create a NOOB MAUI app (MauiApp2). Selected .NET 7 as target. It builds just fine.
- Listing the SDKs installed I noted that my 6.0 sdk had been uninstalled This was the list:

```xml
PS C:\WINDOWS\system32> dotnet --list-sdks
1.1.11 [C:\Program Files\dotnet\sdk]
1.1.12 [C:\Program Files\dotnet\sdk]
1.1.13 [C:\Program Files\dotnet\sdk]
1.1.14 [C:\Program Files\dotnet\sdk]
2.1.202 [C:\Program Files\dotnet\sdk]
2.1.526 [C:\Program Files\dotnet\sdk]
2.1.818 [C:\Program Files\dotnet\sdk]
2.2.401 [C:\Program Files\dotnet\sdk]
2.2.402 [C:\Program Files\dotnet\sdk]
3.1.425 [C:\Program Files\dotnet\sdk]
5.0.203 [C:\Program Files\dotnet\sdk]
5.0.214 [C:\Program Files\dotnet\sdk]
5.0.403 [C:\Program Files\dotnet\sdk]
5.0.406 [C:\Program Files\dotnet\sdk]
5.0.408 [C:\Program Files\dotnet\sdk]
7.0.100 [C:\Program Files\dotnet\sdk]
```
- Used the `dotnet workload install maui-windows maui-android maui-ios maui-maccatalyst --source "https://api.nuget.org/v3/index.json" --sdk-version 6.0.403` command in a PS window to install my missing .NET SDK.
- NOOB .NET 6 project (MauiApp1) still has Android error but no other errors.
- As all my MAUI projects targeted `net6.0-android` I was apparently building with the default SDK 12 (API 31). Opened up the Android SDK manager and noted that the SDK that I was building with was removed. 

![image](https://user-images.githubusercontent.com/5980709/201112020-4ae3df1f-f306-4df4-994d-f29a0bc67557.png)

- Updating to `net6.0-android33.0` corrects the Android API error in NOOB project and my projects. NOOB .NET 6 project will build at this time.
- However, after .NET 6 SDK install I'm still getting the errors with my .NET 6 projects. (as listed above)
- I added Central Package management to the NOOB project and don't get any NU1008 errors.

I'm working from this point to try to correct the build problems. 

These are my install logs from yesterday. At least I believe these are the correct files.
[dd_installer_elevated_20221109074624.zip](https://github.com/dotnet/sdk/files/9981710/dd_installer_elevated_20221109074624.zip)

**Comment by bakerhillpins** [comment]
Following up here as I seem to be working now but I'm relatively sure my install is corrupted. I need to get work done so I'm just going to leave it for now. 

I ended up uninstalling both versions of VS 2022 (pro and pro preview) and just reinstalling Pro. This didn't fix anything. My Android emulators are all botched up and that didn't help that either (had to delete and recreate them). Kinda a mess all around. 

The show missing workloads error seems to root in a completely botched up .NET SDK 6.0.xxx install.  I installed 6.0.403 above and that was successful but didn't solve the issue. I've got a global.json file in the root dir as follows:

```json
{
  "sdk": {
    "version": "6.0.300",
    "rollForward": "latestMinor"
  }
}
```

So I tried to install 6.0.300 with the following:
`dotnet workload install maui-windows maui-android maui-ios maui-maccatalyst --source "https://api.nuget.org/v3/index.json" --sdk-version 6.0.300`
which succeeds but when I list SDKs it's not listed as installed.
```
PS C:\WINDOWS\system32> dotnet --list-sdks
1.1.11 [C:\Program Files\dotnet\sdk]
1.1.12 [C:\Program Files\dotnet\sdk]
1.1.13 [C:\Program Files\dotnet\sdk]
1.1.14 [C:\Program Files\dotnet\sdk]
2.1.202 [C:\Program Files\dotnet\sdk]
2.1.526 [C:\Program Files\dotnet\sdk]
2.1.818 [C:\Program Files\dotnet\sdk]
2.2.401 [C:\Program Files\dotnet\sdk]
2.2.402 [C:\Program Files\dotnet\sdk]
3.1.425 [C:\Program Files\dotnet\sdk]
5.0.203 [C:\Program Files\dotnet\sdk]
5.0.214 [C:\Program Files\dotnet\sdk]
5.0.403 [C:\Program Files\dotnet\sdk]
5.0.406 [C:\Program Files\dotnet\sdk]
5.0.408 [C:\Program Files\dotnet\sdk]
6.0.403 [C:\Program Files\dotnet\sdk]
7.0.100 [C:\Program Files\dotnet\sdk]
PS C:\WINDOWS\system32>
```
The install operation for .300 reports lots of already installed results like below. Not sure why as I expected to be able to do side by side installs.
```
Installing pack Microsoft.Maui.Core.Ref.win version 7.0.49...
Pack Microsoft.Maui.Core.Ref.win version 7.0.49 is already installed.
```

I tried removing the global.json file (I just renamed it)  and it caused the "ShowMissingWorkloads" problem to go away. I'm assuming that it's because it's now targeting the .403 version. I thought that the `"rollForward": "latestMinor"` setting should have forced it to pick up on the .403 version?  

Removing the identical global.json file has worked for multiple projects at this point. So, I'm back to active dev at the moment. Maybe you'll find this useful.
